File: 2023/17/solution.js
const fs = require("fs");
const path = require("path");
const MinHeap = require("../../utils/MinHeap");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("").map(Number));

const turnDirections = [
  [0, 1],
  [1, 0],
  [-1, 0],
  [0, -1],
];

function stateKey(x, y, dir, steps) {
  return `${x},${y},${dir},${steps}`;
}

function findMinPath(queue, visited, minSteps, maxSteps, input) {
  // queue: [][heatLoss, x, y, dirIndex, steps]
  // visited: Map[stateKey] = heatLoss
  // minSteps: minimum number of steps to stay in the same direction
  // maxSteps: maximum number of steps before turning
  // input: grid

  while (queue.length) {
    const [heatLoss, x, y, dirIndex, steps] = MinHeap.pop(queue);
    if (x === input[0].length - 1 && y === input.length - 1) {
      return heatLoss;
    }
    for (let i = 0; i < turnDirections.length; i++) {
      if (steps < minSteps && i != dirIndex) continue; // must stay in the same direction for at least minSteps
      if (i === 3 - dirIndex) continue; // no u-turns)
      if (i === dirIndex && steps === maxSteps) continue; // must turn after maxSteps
      const [dx, dy] = turnDirections[i];
      const [newX, newY] = [x + dx, y + dy];
      if (
        newX < 0 ||
        newX >= input[0].length ||
        newY < 0 ||
        newY >= input.length
      )
        continue;
      const newSteps = i === dirIndex ? steps + 1 : 1;
      const newHeatLoss = heatLoss + input[newY][newX];
      const nextOne = [newHeatLoss, newX, newY, i, newSteps];
      const state = stateKey(newX, newY, i, newSteps);
      if (visited.has(state) && visited.get(state) <= newHeatLoss) continue;
      visited.set(state, newHeatLoss);
      MinHeap.push(queue, nextOne);
    }
  }
  return -1;
}

const part1 = findMinPath([[0, 0, 0, -1, 0]], new Map(), 0, 3, input);
console.log(part1);
const part2 = findMinPath([[0, 0, 0, -1, 5]], new Map(), 4, 10, input);
// use starting steps of 5 to make it choose a new direction
console.log(part2);

/*
  Old solution:
  part1: 6.462s
  part2: 51.564s

  New solution with minHeap:
  part1: 230.407ms
  part2: 744.543ms
*/

File: 2023/17/solution.go
package main

import (
	"container/heap"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

type Value struct {
	heatLoss           int
	pos                [2]int
	dir                [2]int
	stepsSinceLastTurn int
}

type State struct {
	pos                [2]int
	dir                [2]int
	stepsSinceLastTurn int
}

var directions = [][2]int{
	{-1, 0},
	{1, 0},
	{0, -1},
	{0, 1},
}

func parseInput(data []string) [][]int {
	puzzleInput := make([][]int, len(data))
	for i, line := range data {
		row := make([]int, len(line))
		for j, ch := range line {
			row[j] = int(ch - '0')
		}
		puzzleInput[i] = row
	}
	return puzzleInput
}

func distanceBetweenPoints(p1, p2 [2]int) int {
	return utils.AbsInt(p1[0]-p2[0]) + utils.AbsInt(p1[1]-p2[1])
}

func solve(puzzleInput [][]int, minSteps, maxSteps int) int {
	queue := utils.NewMinHeap[Value]()
	heap.Push(queue, &utils.Item[Value]{
		Priority: 0,
		Value:    Value{0, [2]int{0, 0}, [2]int{0, 0}, minSteps},
	})
	visited := make(map[State]int)

	rMax := len(puzzleInput) - 1
	cMax := len(puzzleInput[0]) - 1
	target := [2]int{rMax, cMax}
	var r, c, r2, c2, heatLoss, steps int
	var pos, dir [2]int
	var state State

	for len(queue.PriorityQueue) > 0 {
		item := heap.Pop(queue).(*utils.Item[Value])
		if item.Value.pos == target {
			return item.Value.heatLoss
		}
		r, c = item.Value.pos[0], item.Value.pos[1]
		dir = item.Value.dir
		for _, d := range directions {
			if d[0] == -dir[0] && d[1] == -dir[1] {
				continue
			}
			r2, c2 = r+d[0], c+d[1]
			if r2 < 0 || c2 < 0 || r2 > rMax || c2 > cMax {
				continue
			}
			steps = 1
			if d[0] == dir[0] && d[1] == dir[1] {
				steps = item.Value.stepsSinceLastTurn + 1
			} else if item.Value.stepsSinceLastTurn < minSteps {
				continue
			}
			if steps > maxSteps {
				continue
			}
			pos = [2]int{r2, c2}
			heatLoss = item.Value.heatLoss + puzzleInput[r2][c2]
			state = State{pos, d, steps}
			if h, ok := visited[state]; ok && h <= heatLoss {
				continue
			}
			visited[state] = heatLoss
			heap.Push(queue,
				&utils.Item[Value]{
					Priority: heatLoss + distanceBetweenPoints(pos, target),
					Value:    Value{heatLoss, pos, d, steps},
				})
		}
	}
	return -1
}

func part1(puzzleInput [][]int) int {
	return solve(puzzleInput, 0, 3)
}

func part2(puzzleInput [][]int) int {
	return solve(puzzleInput, 4, 10)
}
